### PR TITLE
Automatically assemble app + tests on runFlank

### PR DESCRIPTION
**Pull request by asadsalman**
This PR adds a new property, `dependOnAssemble`, in the `fladle` extension. This boolean can be used to turn on automatic building of app + test apk on `execFlank` and `runFlank`.

This is useful because now fladle can execute as soon as a project's APKs are available. As opposed to running `./gradlew assemble assembleAndroidTest` and then running fladle, where we would have to wait for _all_ APKs to be built before we can run fladle.

### CODE REVIEW

File: fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt [hunk 1]
```diff
@@ -106,6 +107,10 @@ class FladlePluginDelegate {
         environment(mapOf("GOOGLE_APPLICATION_CREDENTIALS" to config.serviceAccountCredentials.get()))
       }
       dependsOn(writeConfigProps)
+      if (config.dependOnAssemble.isPresent && config.dependOnAssemble.get()) {
+        dependsOn(ASSEMBLE_ANDROID_TEST)
+        dependsOn("assemble")
```
Comment (runningcode) [inline fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt:112]: I don't think this should depend on assemble. That would build both debug and release variants. In this case we likely only want the debug variant.

We also want to make sure that this works for all variants. Can you please add a test which tests that the proper task dependencies are executed?

**Review by runningcode — COMMENTED** [review]
Thanks for the contribution, let's make sure this works in edge cases like different build flavors and add some testing.

File: fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt (hunk 1 above)
Comment (asadsalman) [reply fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt]: Using `assembleProvider` to get appropriate assemble task now. Will add test!

**Comment by asadsalman** [comment]
Once this is approved/merged, I can send in a PR to `gh-pages` branch to update docs.

**Comment by runningcode** [comment]
@asadsalman I appreciate that, but there is no need to. The docs are updated automatically based on the contents in `docs`.
Please update the documentation in this PR.

File: fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt [hunk 2]
```diff
@@ -106,6 +108,15 @@ class FladlePluginDelegate {
...
       }
       dependsOn(writeConfigProps)
+      if (config.dependOnAssemble.isPresent && config.dependOnAssemble.get()) {
+        val testedExtension = project.extensions.findByType(TestedExtension::class.java)
+        testedExtension?.testVariants?.configureEach {
+          if (testedVariant.assembleProvider.isPresent) {
+            dependsOn(testedVariant.assembleProvider.get())
+          }
+        }
+        dependsOn(ASSEMBLE_ANDROID_TEST)
```
Comment (runningcode) [inline fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt:118]: why is this a contant? Couldn't there be multiple test variants as well?

File: fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt [hunk 3]
```diff
@@ -106,6 +108,15 @@ class FladlePluginDelegate {
         environment(mapOf("GOOGLE_APPLICATION_CREDENTIALS" to config.serviceAccountCredentials.get()))
       }
       dependsOn(writeConfigProps)
+      if (config.dependOnAssemble.isPresent && config.dependOnAssemble.get()) {
+        val testedExtension = project.extensions.findByType(TestedExtension::class.java)
+        testedExtension?.testVariants?.configureEach {
+          if (testedVariant.assembleProvider.isPresent) {
+            dependsOn(testedVariant.assembleProvider.get())
```
Comment (runningcode) [inline fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt:115]: nice, thanks for finding and fixing this!

**Review by runningcode — COMMENTED** [review]
looks good! just need tests and doc update!

File: fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt (hunk 3 above)
Comment (asadsalman) [reply fladle-plugin/src/main/java/com/osacky/flank/gradle/FladlePluginDelegate.kt]: Happy to help! 😃

File: fladle-plugin/src/main/java/com/osacky/flank/gradle/FladleConfig.kt [hunk 4]
```diff
@@ -427,6 +427,14 @@ interface FladleConfig {
   @get:Optional
   val additionalFlankOptions: Property<String>
 
+  /**
+   * Boolean to allow assembling application and test code automatically
```
Comment (runningcode) [inline fladle-plugin/src/main/java/com/osacky/flank/gradle/FladleConfig.kt:431]:
```suggestion
   * When enabled, the execution of flank will depend on the Gradle tasks to assemble the debug and instrumentation APKs.
```

File: fladle-plugin/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt [hunk 5]
```diff
@@ -29,7 +29,7 @@ open class FlankGradleExtension @Inject constructor(objects: ObjectFactory) : Fl
   override val useOrchestrator: Property<Boolean> = objects.property<Boolean>().convention(false)
   override val autoGoogleLogin: Property<Boolean> = objects.property<Boolean>().convention(false)
   override val devices: ListProperty<Map<String, String>> = objects.listProperty<Map<String, String>>().convention(listOf(mapOf("model" to "NexusLowRes", "version" to "28")))
-
+  override val dependOnAssemble: Property<Boolean> = objects.property<Boolean>().convention(false)
```
Comment (runningcode) [inline fladle-plugin/src/main/java/com/osacky/flank/gradle/FlankGradleExtension.kt:32]: can this be added at the end of the list beneath `additionalGcloudOptions`? so that it is consistent with the API documentations

File: fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt [hunk 6]
```diff
@@ -126,6 +127,69 @@ class FlankGradlePluginIntegrationTest {
...
+      testProjectRoot.newFile("foo").writeText("{}")
+      testProjectRoot.newFolder("src/main")
+      testProjectRoot.newFile("src/main/AndroidManifest.xml").writeText("""
+          <?xml version="1.0" encoding="utf-8"?>
+          <manifest package="com.osacky.flank.gradle.sample" xmlns:android="http://schemas.android.com/apk/res/android" />
+      """.trimIndent())
+      return GradleRunner.create()
+          .withProjectDir(testProjectRoot.root)
+          .withPluginClasspath()
+          .withGradleVersion("6.1.1")
```
Comment (runningcode) [inline fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt:169]: let's not specify a Gradle version and just use the same one as the build. this will prevent downloading too many Gradle versions.

File: fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt [hunk 7]
```diff
@@ -126,6 +127,69 @@ class FlankGradlePluginIntegrationTest {
...
+          .withArguments("runFlank")
+          .buildAndFail()
```
Comment (runningcode) [inline fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt:171]: can we verify that the error message is exactly what we expect it to be here? there could be other reasons for the build failure here and we just want to make sure it is the same one as what we expected here.

File: fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt [hunk 8]
```diff
@@ -126,6 +127,69 @@ class FlankGradlePluginIntegrationTest {
...
+  }
+
+  @Test
+  fun testWithDependOnAssemble() {
+      val result =  setUpDependOnAssemble(true)
+      result.tasks.filter { it.path.startsWith(":assembleDebug") }.map { it.path }.let { tasks ->
```
Comment (runningcode) [inline fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt:177]:
```suggestion
        assertThat(result.task(":assembleDebug")!!.outcome).isEqualTo(TaskOutcome.SUCCESS)
        assertThat(result.task(":assembleDebugAndroidTest")!!.outcome).isEqualTo(TaskOutcome.SUCCESS)
```

**Review by runningcode — COMMENTED** [review]
looking good. just some minor nits!

File: fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt (hunk 6 above)
Comment (asadsalman) [reply fladle-plugin/src/test/java/com/osacky/flank/gradle/integration/FlankGradlePluginIntegrationTest.kt]: It looks like using anything less than 6.1.1 fails.
image

**Review by runningcode — APPROVED** [review]

**Comment by runningcode** [comment]
Thanks!

**Comment by asadsalman** [comment]
🎊
